### PR TITLE
Min-cut partitioner always saves tensors that are returned as-is in backward

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119708
* __->__ #114970

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -836,8 +836,18 @@ def get_node_weight(node) -> int:
...
+                continue
+            # If someone saves a input for backward as-is and backward
+            # returns that tensor as-is as a grad input, then the node x would
+            # be both a required_bw_node and an input. In this case we
+            # (1) connect x_in to to the source, (2) x_out to the sink, and
+            # (3) assign the proper weight to the x_in-x_out edge, so that
+            # x would be part of cut nodes. A case where this happens is if
+            # NestedTensor saves a offset tensor as part of the singleton int
+            # in sizes.
+            nx_graph.add_edge(node.name + "_out", "sink", capacity=math.inf)
```
Comment (soulitzer) [inline torch/_functorch/partitioners.py:850]: Alternatively we could just update the cut-set manually

Comment (Chillee) [reply torch/_functorch/partitioners.py:850]: Could you add a test?

Comment (soulitzer) [reply torch/_functorch/partitioners.py:850]: Added

**Comment by jansel** [comment]
@Chillee should review this one

**Review by Chillee — APPROVED** [review]
What happens today with this? An error because it can't find a valid cut?

**Comment by soulitzer** [comment]
Basically yeah, when we go to extract the modules corresponding to the input/outputs, it would error out because one of outputs of the graph is not passed as in as input anywhere.

**Pull request closed without merging**